### ISSUE TITLE
Add Request Interceptor

### ISSUE THREAD

**Issue by BrandonRomano** [BrandonRomano/wrecker#4]
Add a request interceptor to the Wrecker struct.

Example use case:

An API has an `API_KEY` that has to be passed in every request.  It would be very tedious to call `WithParam(API_KEY)` on every request.  Rather it would be nice to set this on the Wrecker instance once + all requests are updated before being sent out.

**Comment by benpate** [comment]
I was going to make a separate issue, but I think it would probably be covered by this ticket, so I'll just add another use case to the discussion.

It would be good to have some kind of debugging output for when you're trying to troubleshoot a particular request.  Instances like "is it my fault, or is it the servers?"  I think a nicely designed request interceptor could probably cover this, but it would probably help to have this in mind while it's being created.

**Comment by BrandonRomano** [comment]
Hey @benpate.

I'm actually not quite sure I fully understand your use case.

In my mind, a request interceptor is something that modify a request right before it's sent out.  So the request interceptor would never have access to the actual response from the server.

What I think that you want is going to be coming from this PR #9 once it's finished.  After that PR is merged in, there will be 2 values being returned from requests:  the response + an error.

An error is going to be the thing that tells you that you've done something wrong (if it's not nil).  The request is going to have all of the information from the server, which would have information like the StatusCode (which could tell you that the server did something wrong).

So logging can be added (and probably should, I'll make an issue for it), but this would not be related to the actual request interceptor.

**Comment by benpate** [comment]
Gotcha -- and thanks for your attention on this.  I didn't say it before, but good job on making this library.  I think it's really going to help me out.

**Comment by BrandonRomano** [comment]
Thanks!  And thank you for contributing!

**Comment by BrandonRomano** [comment]
@benpate assigning this to you, as you mentioned you were working on this in a remote branch!

**Comment by benpate** [comment]
Hey.  I just made a new branch, and will be getting code posted shortly.  With all the API changes, it's probably best to single-thread them for now, so I've based the new branch off of the other pull requests I'm working on.

One question.  I'd originally mocked up BEFORE and AFTER interceptors -- so the BEFORE interceptor would work on the Request, and the AFTER interceptor would work on the response.  Does this make sense to you?  Or, since we're now returning the response directly, would you rather just have the BEFORE interceptor and be done?

**Comment by BrandonRomano** [comment]
Hmm.  So the "After" interceptor would be more of a `Response Interceptor`, where as the before would be the `Request Interceptor`.  I would be interested in hearing more about the After / Response Interceptor, but I would say for the initial PR let's keep it small and only do the Request Interceptor.

**Comment by benpate** [comment]
One more thought about this.  [AngularJS](https://docs.angularjs.org/api/ng/service/$http) has put a lot of thought into how this works, and I think we could benefit from modelling them..  It handles Interceptors as a collection of functions that are executed both before and after a request.  So, they're bundling request and response interceptors together into a single structure.  

This seems like it could be cool, because then you can distribute a single Interceptor that would work as a complete package.

**Comment by BrandonRomano** [comment]
> TLDR: I think I agree with you if my assumptions are correct... Just verify that I'm correct in this:

Gotcha.  I see what you mean here... I think we're actually getting tripped up on verbiage, but actually are very close being in line in line with each other.

In Angular, these are generically referred to as `Interceptors`.  Here's an Angular example (simplified to match our use case):

``` js
$provide.factory('myHttpInterceptor', function($q, dependency1, dependency2) {
  return {

    // Optional
    'request': function(config) {     // This is the thing I am calling a RequestInterceptor
      return config;
    },

    // Optional
    'response': function(response) {   // This is the thing I am calling the ResponseInterceptor
      return response;
    }

  };
});
```

I see what you're saying now and I do appreciate how Angular handles this.

Here's some go code so I can kind of see if I understand the direction you want to take this:

``` go
type RequestInterceptor func(*wrecker.Request)
type ResponseInterceptor func(*http.Response)

type Interceptor struct {
    request     RequestInterceptor      // Optional
    response  ResponseInterceptor   // Optional
}
```

And then you would create an instance of an `Interceptor`, and do:

``` go
wreckerClient = &wrecker.Wrecker{
    BaseURL: "http://localhost:5000",
    HttpClient: &http.Client{
        Timeout: 10 * time.Second,
    },
    DefaultContentType: "application/x-www-form-urlencoded",
    Interceptor: &Interceptor{
        request: myRequestInterceptorFunction,
        response: myResponseInterceptorFunction,
    },
}
```

Obviously not 100% accurate, but this is the general idea correct?

---

So what I was thinking was that we would do a similar thing except they be different values.

``` go
type RequestInterceptor func(*wrecker.Request)
type ResponseInterceptor func(*http.Response)

wreckerClient = &wrecker.Wrecker{
    BaseURL: "http://localhost:5000",
    HttpClient: &http.Client{
        Timeout: 10 * time.Second,
    },
    DefaultContentType: "application/x-www-form-urlencoded",
    RequestInterceptor: myRequestInterceptorFunction,
    ResponseInterceptor: myResponseInterceptorFunction
}
```

---

Now that I'm looking at both, I totally agree with you and I prefer your pattern.

**Comment by benpate** [comment]
Awesome.  For the record, this just dawned on me this morning, when I was putting this code in place.  

I still agree with your original idea to get one piece in at a time.  Let me know if you'd like to start with this first, and then go back later to add in the response interceptor, or just keep pushing forward with the code you posted here.

**Comment by BrandonRomano** [comment]
Definitely, yeah -- I think the best bet would be to add the request interceptor in 1 PR, and then tack on the response interceptor in another.  They can be very close together, but I think it would be beneficial to have a PR specifically for each one (for readability + referenceability purposes).

So I suggest for the first PR, this would be the extent of the interceptor object:

``` go
wreckerClient = &wrecker.Wrecker{
    BaseURL: "http://localhost:5000",
    HttpClient: &http.Client{
        Timeout: 10 * time.Second,
    },
    DefaultContentType: "application/x-www-form-urlencoded",
    Interceptor: &Interceptor{
        request: myRequestInterceptorFunction,
    },
}
```

**Comment by benpate** [comment]
I've been doing some thinking about this, and there's a core issue that we (really you) need to decide: **When should the request interceptor run?**  

Currently, it runs on the wrecker.Request data BEFORE it is converted into an HTTPRequest.  That makes the API very clean, because everything is operating on Wrecker objects.

But, I've run into more than one case where it might be better to operate on the HTTPRequest itself -- AFTER Wrecker has done its work.  Having request interceptors work on the native objects would enable plugins like GZip compression to be implemented as interceptors, instead of requiring another mechanism.

What do you think, Boss?

**Comment by BrandonRomano** [comment]
Fantastic thought -- Why not both :tada: ?

I say the Before request would be the one we should implement first (and also call `Request`).  This is the one I think will be 90%+ of use cases.

I'm not sure of a good name for the one that happens after the `HTTPRequest` conversion.  Perhaps you have a suggestion?

**Comment by benpate** [comment]
Sure.  I think that will work.  I'll knock it around a bit and then post a suggestion.

**Comment by benpate** [comment]
I've just posted updates to PR #24 that supports Interceptors with functions at three different stages of the request: 

1: **WreckerRequest** receives the completed wrecker.Request object before it's converted into HTTP
2: **HTTPRequest** receives the completed http.Request object before it's sent to the remote server.
3: **HTTPResponse** receives the http.Response object and body before it's parsed and returned to the caller.

There's also a sample `Debug` interceptor that is really helping me out.  It may even satisfy Issue #10 when it's complete.

I still need to include several more test cases and perhaps another sample Interceptor.  But, I think this is going really well.  Can you please give me some feedback to make sure I'm on the right track?
